copy/multiple.go

```diff
@@ -27,23 +32,90 @@ const (
 type instanceCopy struct {
 	op           instanceCopyKind
 	sourceDigest digest.Digest
+
+	// Fields which can be used by callers when operation
+	// is `instanceCopyClone`
+	compressionVariant            *compression.Algorithm
+	cloneAddPlatform              *imgspecv1.Platform
+	cloneAddAnnotations           map[string]string
+	cloneAddCompressionAlgorithms []compression.Algorithm
+}
+
+// internal type only to make imgspecv1.Platform comparable
+type platformComparable struct {
+	Architecture string
+	OS           string
+	OSVersion    string
+	OSFeatures   string
+	Variant      string
+}
+
+// Converts imgspecv1.Platform to a comparable format.
+func platformV1ToPlatformComparable(platform imgspecv1.Platform) platformComparable {
+	return platformComparable{Architecture: platform.Architecture,
+		OS:         platform.OS,
+		OSFeatures: strings.Join(platform.OSFeatures, ","),
+		OSVersion:  platform.OSVersion,
+		Variant:    platform.Variant,
+	}
+}
+
+// platformCompresionPair prepares a mapping of platformComparable -> CompressionAlgorithmNames for given digests
+func platformCompresionPair(list internalManifest.List, instanceDigests []digest.Digest) (map[platformComparable]*set.Set[string], error) {
+	platformCompressionMap := make(map[platformComparable]*set.Set[string])
+	for _, instanceDigest := range instanceDigests {
+		instanceDetails, err := list.Instance(instanceDigest)
+		if err != nil {
+			return nil, fmt.Errorf("getting details for instance %s: %w", instanceDigest, err)
+		}
+		platform := platformV1ToPlatformComparable(*instanceDetails.ReadOnly.Platform)
+		_, valueExists := platformCompressionMap[platform]
+		if !valueExists {
+			platformCompressionMap[platform] = set.New[string]()
+		}
+		set := platformCompressionMap[platform]
+		set.AddSlice(instanceDetails.ReadOnly.CompressionAlgorithmNames)
+	}
+	return platformCompressionMap, nil
 }
 
 // prepareInstanceCopies prepares a list of instances which needs to copied to the manifest list.
-func prepareInstanceCopies(instanceDigests []digest.Digest, options *Options) []instanceCopy {
+func prepareInstanceCopies(list internalManifest.List, instanceDigests []digest.Digest, options *Options) ([]instanceCopy, error) {
 	res := []instanceCopy{}
+	platformCompressionMap, err := platformCompresionPair(list, instanceDigests)
+	if err != nil {
+		return res, err
+	}
 	for i, instanceDigest := range instanceDigests {
 		if options.ImageListSelection == CopySpecificImages &&
```

Comment:
Combining `CopySpecificImages` and the clone map is yet one more thing to worry about.

I think the current approach is correct in that if `ImageListSelection` excludes an image, we should not be cloning it.

But what if `ImageListSelection` refers to a gzip image, we are asked to ensure Zstd exists, and `list` already contains a Zstd variant… but not in `ImageListSelection`?

I think one very reasonable option is to just fail on `ImageListSelection && EnsureCompression…` with a “not currently implemented” error. And we can implement that later if a user shows up. (As a workaround, the user could do two copies, one with each option; with the order depending on what they want to achieve.)

Reply:
Done, erroring out for this and added test to verify the same.